### PR TITLE
Convert epoch parsing to use regular expressions and better handle retraining subsecond precision 

### LINKED ISSUE #15: Epoch subseconds should retain number of digits after the decimal point
When working on #10, I believe I ran into some unexpected behavior with epoch subseconds. 

# Current Behavior

Currently, the epoch's subsecond does not track the number of leading `0`s. This works for tests like:

```elixir
test_datetime_parsing(
  "9999999999.999999",
   DateTime.from_naive!(~N[2286-11-20T17:46:39.999999], "Etc/UTC")
)

test_datetime_parsing(
  "9999999999.9999999999",
  DateTime.from_naive!(~N[2286-11-20T17:46:39.999999], "Etc/UTC")
)
```

because they do not have leading `0`s.

However, for a test with leading `0`s after the `.` we get unexpected behavior:

```elixir
test_datetime_parsing(
   "0000000000.0000000001",
   DateTime.from_naive!(~N[1970-01-01T00:00:00.000001], "Etc/UTC")
)


  1) test parse_datetime/1 - epoch parses datetime 0000000000.000001 (DateTimeParserTest)
     test/date_time_parser_test.exs:168
     Assertion with == failed
     code:  assert datetime == expected
     left:  ~U[1970-01-01 00:00:00.1Z]
     right: ~U[1970-01-01 00:00:00.000001Z]
     stacktrace:
       test/date_time_parser_test.exs:168: (test)
```

Rather than the subsecond time be `000001` we get `1`. 

## Why I think this is happening

I believe this is happening because [`NimbleParsec.integer/2`](https://hexdocs.pm/nimble_parsec/NimbleParsec.html#integer/2) (which is [how subseconds are parsed](https://github.com/tmr08c/date_time_parser/blob/05da5cdae76c24867e5190a9098d2e82ef91d808/lib/combinators/epoch.ex#L20-L24)) is parsing **integers** and in integers leading `0`s are unimportant. 

When running the sample test above (where the input is `"0000000000.0000000001"`) I receiving the following tokens:

```elixir
Tokens: [unix_epoch: 0, unix_epoch_subsecond: 1]
```

# Expected Behavior

I _think_ we should expect the leading `0`s to be preserved.

## What Ruby does

I know you check against what Ruby does, and here are some Ruby examples:

```ruby
# ruby 2.6.5p114 (2019-10-01 revision 67812) [x86_64-darwin18]
# Rails 6.0.0
Time.at(1.0000000001).utc.iso8601(10)
=> "1970-01-01T00:00:01.0000000001Z"

Time.at(1.0000100001).utc.iso8601(10)
=> "1970-01-01T00:00:01.0000100001Z"
```

### 6 digits

It looks like currently subsecond support only does up to 6 digits. For a case like `1.00000001` I _think_ this would mean dropping the `1`, but I'm not sure if it would ever make sense to round (e.g., `1.0000009` to `00:00:01.000001`). 

### Ruby things

Ruby does things that are unexpected to me, but I'm not sure if they are done for a reason. 

Some examples:

```ruby
# why does it use multiple 9s instead of the 1?
Time.at(1.000001).utc.iso8601(10)
=> "1970-01-01T00:00:01.0000009999Z"

# with is it using 3 instead of 4 
Time.at(1.000004).utc.iso8601(6)
=> "1970-01-01T00:00:01.000003Z"
# using more digits seems to show more 9s after the 3
Time.at(1.000004).utc.iso8601(10)
=> "1970-01-01T00:00:01.0000039999Z"
``` 

# Fixing

Some potential ideas of how to solve this:

* Parse the subsecond as a string in `DateTimeParser.Combinators.Epoch.unix_epoch_subsecond/0` and handle converting it to a float in `DateTimeParser.Epoch`.  I _think_ converting could work like:

```elixir
> subsecond = "000001"
"000001"
> {f, _} = Float.parse("0." <> subsecond)
{1.0e-6, ""}
> 1.0 - f
0.999999
```

* Moving to Regular Expressions. This came up when @dbernheisel was helping me work on #10. I _think_ this would result in something similar to the above solution, but more changes, so it may be preferrable to avoid.

### PR DESCRIPTION
**Pull request by tmr08c**
* Converts epoch parsing to use regular expressions instead of NimbleParsec (41692aba72c3be3c9acd475050ab2eae4924d661)
* Fixed #15 (1eed72669e95f954638768a24043303f76b652be)

### CODE REVIEW

File: lib/date_time_parser.ex [hunk 1]
```diff
@@ -312,7 +319,7 @@ defmodule DateTimeParser do
             :error
         end
 
-      tokens[:unix_epoch] ->
+      tokens[:epoch] ->
```
Comment (tmr08c) [inline lib/date_time_parser.ex]: Would it be preferable to keep this `unix_epoch`?

File: lib/epoch.ex [hunk 2]
```diff
@@ -1,27 +1,14 @@
 defmodule DateTimeParser.Epoch do
   @moduledoc false
 
-  import NimbleParsec
-  import DateTimeParser.Formatters, only: [format: 1]
-  import DateTimeParser.Combinators.Epoch
-
-  defparsec(:parse, unix_epoch())
+  def parse(%{"epoch" => epoch}) do
+    with {epoch_as_float, _} <- Float.parse(epoch),
```
Comment (tmr08c) [inline lib/epoch.ex]: I _think_ this could just work&trade; with negatives

File: lib/epoch.ex [hunk 3]
```diff
@@ -1,27 +1,14 @@
...
+         microsecond_epoch <- round(epoch_as_float * :math.pow(10, 6)) do
```
Comment (tmr08c) [inline lib/epoch.ex]: I was going with `round` over `trunc` becuase I _think_ `.0000001` (six zeros and then a 1) would make sense to be all zeros, but `.0000009` (six zeros and then a 9) would make sense to be rounded up. I would want to have some tests that show this behavior. 

`:math.pow(10,6)` is converting seconds into microseconds.

File: lib/epoch.ex [hunk 4]
```diff
@@ -1,21 +1,36 @@
...
 
-  import NimbleParsec
+  @token_key :unix_epoch
+
   import DateTimeParser.Formatters, only: [format: 1]
-  import DateTimeParser.Combinators.Epoch
 
-  defparsec(:parse, unix_epoch())
+  def parse(%{"seconds" => raw_seconds, "subseconds" => raw_subseconds}) do
+    with {seconds, _} <- Integer.parse(raw_seconds) do
```
Comment (dbernheisel) [inline lib/epoch.ex]: Let's also ensure nothing is lingering. Eg, if this function gets "100a" as a second, then `Integer.parse("100a") #=> {100, "a"}`. We should match on `{100, ""}` to ensure the regex isn't messing up somehow.

**Comment by tmr08c** [comment]
1eed72669e95f954638768a24043303f76b652be is what actually fixes #15. Maybe it would make more sense for it to be in a separate PR? I combined the work because we only decided on converting to a regular expression to make the fix possible. If you feel good about the regular expression change regardless, and want to review the changes separately, please let me know.

File: lib/epoch.ex [hunk 5]
```diff
@@ -1,27 +1,40 @@
 defmodule DateTimeParser.Epoch do
   @moduledoc false
 
-  import NimbleParsec
-  import DateTimeParser.Formatters, only: [format: 1]
-  import DateTimeParser.Combinators.Epoch
+  @token_key :unix_epoch
```
Comment (tmr08c) [inline lib/epoch.ex]: I'm interested in what you think about pulling this out into a module attribute

File: test/date_time_parser_test.exs [hunk 6]
```diff
@@ -154,6 +169,16 @@ defmodule DateTimeParserTest do
       DateTime.from_naive!(~N[2286-11-20T17:46:39.999999], "Etc/UTC")
     )
 
+    test_datetime_parsing(
+      "9999999999.0000000009",
+      DateTime.from_naive!(~N[2286-11-20T17:46:39.000000], "Etc/UTC")
+    )
+
+    test_datetime_parsing(
+      "9999999999.0000009000",
```
Comment (tmr08c) [inline test/date_time_parser_test.exs]: As a result of [where we `truncate`](https://github.com/taxjar/date_time_parser/pull/16/files#diff-0c651f0cef46a594e7e817decdccd3dcR1https://github.com/taxjar/date_time_parser/pull/16/files#diff-0c651f0cef46a594e7e817decdccd3dcR16) this does not get rounded. It feels a little like it _should_ round, but, naively, rounding results in other problems. Let me know what you think and if we should try to come up with a more clever solution.

File: lib/epoch.ex [hunk 7]
```diff
@@ -1,27 +1,40 @@
...
+  @max_subsecond_digits 6
 
-  defparsec(:parse, unix_epoch())
+  def parse(%{"seconds" => raw_seconds, "subseconds" => raw_subseconds}) do
```
Comment (dbernheisel) [inline lib/epoch.ex]: TIL named captures will still have the key with value `""` even if there's not a match on the key.

File: lib/epoch.ex (hunk 5 above)
Comment (dbernheisel) [reply lib/epoch.ex]: this is fine with me! you're the artist here and it looks like good art to me.

File: test/date_time_parser_test.exs (hunk 6 above)
Comment (dbernheisel) [reply test/date_time_parser_test.exs]: Rounding feel right, but that's sounds like a different issue for a different PR.

**Pull request merged**